Return 400 for invalid POST params

composerinteralia opened the pull request:
Closes #40036

Prior to this commit POST requests with invalid parameters did not
consistently return 400 responses. In the case of a request with a
`_method` override, any invalid parameters prevent the [rack method
override middleware][] from determining the method override. At that
point the request comes into Rails as a POST request and the response
status depends on whether another route happens to match the POST
request.

In the case where a route happens to match the POST request, Rails
attempts to serve the request and then fails with
`ActionController::BadRequest` when referencing the invalid [request
parameters][]. This turns into a 400 response.

But if no route matches the POST request, Rails never gets far enough to
reference the invalid request parameters. Instead, the request fails
with `ActionController::RoutingError`, turning into a 404 response.
This is confusing, since it is a 404 for different route than we would
have expected to hit, given the `_method` override.

This commit ensures that we respond with a 400 status regardless of
whether an unrelated route happens to match. It does this by referencing
the parameters before passing the request off to Journey. If they are
invalid, referencing them will raise the `ActionController::BadRequest`,
triggering the 400 response.

Because this commit references the parameters earlier than before, it
also needs to set "rack.input" in the `ActionDispatch::RoutingVerbs`
test class to keep the tests green. In a real application I believe
"rack.input" should have already been set by the handler (see [Puma][],
for example).

[rack method override middleware]: https://github.com/rack/rack/blob/249dd785625f0cbe617d3144401de90ecf77025a/lib/rack/method_override.rb
[request parameters]: https://github.com/rails/rails/blob/7af59e16a255da5aee2863af0eefaa9f0cfd857d/actionpack/lib/action_dispatch/http/parameters.rb#L50
[Puma]: https://github.com/puma/puma/blob/1b302a5de07b974f08b20b87c86326ee0e57271d/lib/puma/server.rb#L557

p8 commented on actionpack/test/dispatch/routing_test.rb:
```diff
@@ -4609,6 +4613,17 @@ def show
+        patch "/foo/:id", to: "test_invalid_urls/foo#update"
+      end
+
+      post "/foo/1", params: "_method=PATCH&key[]=value&key[0]=value"
```
It's a bit unclear from the PR what is meant by invalid params.
I had to read the original issue to find out it is a param that is both an array and a hash.
```suggestion
      invalid_params = "key[]=value&key[0]=value"
      post "/foo/1", params: "_method=PATCH&#{invalid_params}"
```

composerinteralia replied:
Thanks! Your suggestion it definitely an improvement.

`invalid_params` could basically be anything that `Rack::QueryParser` is unable to parse. It doesn't necessarily have to be an array and a hash. I am going to simplify this to "%", which is a minimal invalid parameter.

p8 replied:
That's even better 😄

composerinteralia committed:
Return 400 for invalid POST params

Closes https://github.com/rails/rails/issues/40036

Prior to this commit POST requests with invalid parameters did not
consistently return 400 responses. In the case of a request with a
`_method` override, any invalid parameters prevent the [rack method
override middleware][] from determining the method override. At that
point the request comes into Rails as a POST request and the response
status depends on whether another route happens to match the POST
request.

In the case where a route happens to match the POST request, Rails
attempts to serve the request and then fails with
`ActionController::BadRequest` when referencing the invalid [request
parameters][]. This turns into a 400 response.

But if no route matches the POST request, Rails never gets far enough to
reference the invalid request parameters. Instead, the request fails
with `ActionController::RoutingError`, turning into a 404 response.
This is confusing, since it is a 404 for different route than we would
have expected to hit, given the `_method` override.

This commit ensures that we respond with a 400 status regardless of
whether an unrelated route happens to match. It does this by referencing
the parameters before passing the request off to Journey. If they are
invalid, referencing them will raise the `ActionController::BadRequest`,
triggering the 400 response.

Because this commit references the parameters earlier than before, it
also needs to set "rack.input" in a couple of test classes to keep the
tests green. In a real application I believe "rack.input" should have
already been set by the handler (see [Puma][], for example).

[rack method override middleware]: https://github.com/rack/rack/blob/249dd785625f0cbe617d3144401de90ecf77025a/lib/rack/method_override.rb
[request parameters]: https://github.com/rails/rails/blob/7af59e16a255da5aee2863af0eefaa9f0cfd857d/actionpack/lib/action_dispatch/http/parameters.rb#L50
[Puma]: https://github.com/puma/puma/blob/1b302a5de07b974f08b20b87c86326ee0e57271d/lib/puma/server.rb#L557

p8 commented on actionpack/lib/action_dispatch/routing/route_set.rb:
```diff
@@ -840,6 +840,7 @@ def url_for(options, route_name = nil, url_strategy = UNKNOWN, method_name = nil
 
       def call(env)
         req = make_request(env)
+        req.parameters
```
Calling `parameters` earlier should probably not impact performance as the results get calculated once and are memoized in the header.

composerinteralia replied:
Yeah, that was my thinking as well.

seanpdoyle commented:
@composerinteralia Rails bot has recently closed the original issue (https://github.com/rails/rails/issues/40036). Is this still worth pursuing?

composerinteralia commented:
Thanks Sean. This is pretty stale, so I'm going to close for now.
